plugins/backstage-plugin-coder/src/utils/StateSnapshotManager.ts

```diff
@@ -0,0 +1,166 @@
+  // with -0 and +0. Still need === to handle that comparison
+  return Object.is(v1, v2) || (v1 === 0 && v2 === 0);
+}
+
+/**
+ * Favors shallow-ish comparisons (will check one level deep for objects and
+ * arrays, but no more)
+ */
+export function defaultDidSnapshotsChange<TSnapshot extends ReadonlyJsonValue>(
+  oldSnapshot: TSnapshot,
+  newSnapshot: TSnapshot,
+): boolean {
+  if (areSameByReference(oldSnapshot, newSnapshot)) {
+    return false;
+  }
+
+  const oldIsPrimitive =
+    typeof oldSnapshot !== 'object' || oldSnapshot === null;
+  const newIsPrimitive =
+    typeof newSnapshot !== 'object' || newSnapshot === null;
+
+  if (oldIsPrimitive && newIsPrimitive) {
+    const numbersAreWithinTolerance =
+      typeof oldSnapshot === 'number' &&
+      typeof newSnapshot === 'number' &&
+      Math.abs(oldSnapshot - newSnapshot) < 0.00005;
+
+    if (numbersAreWithinTolerance) {
+      return false;
+    }
+
+    return oldSnapshot !== newSnapshot;
+  }
+
+  const changedFromObjectToPrimitive = !oldIsPrimitive && newIsPrimitive;
+  const changedFromPrimitiveToObject = oldIsPrimitive && !newIsPrimitive;
+
+  if (changedFromObjectToPrimitive || changedFromPrimitiveToObject) {
+    return true;
+  }
+
+  if (Array.isArray(oldSnapshot) && Array.isArray(newSnapshot)) {
+    const sameByShallowComparison =
+      oldSnapshot.length === newSnapshot.length &&
+      oldSnapshot.every((element, index) =>
+        areSameByReference(element, newSnapshot[index]),
+      );
+
+    return !sameByShallowComparison;
+  }
+
+  const oldInnerValues: unknown[] = Object.values(oldSnapshot as Object);
+  const newInnerValues: unknown[] = Object.values(newSnapshot as Object);
+
+  if (oldInnerValues.length !== newInnerValues.length) {
+    return true;
+  }
+
+  for (const [index, value] of oldInnerValues.entries()) {
+    if (value !== newInnerValues[index]) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * @todo Might eventually make sense to give the class the ability to merge
+ * snapshots more surgically and maximize structural sharing (which should be
+ * safe since the snapshots are immutable). But we can worry about that when it
+ * actually becomes a performance issue
+ */
+export class StateSnapshotManager<
+  TSnapshot extends ReadonlyJsonValue = ReadonlyJsonValue,
+> implements SnapshotManagerApi<TSnapshot>
+{
+  private subscriptions: Set<SubscriptionCallback<TSnapshot>>;
+  private didSnapshotsChange: DidSnapshotsChange<TSnapshot>;
+  private activeSnapshot: TSnapshot;
+
+  constructor(options: SnapshotManagerOptions<TSnapshot>) {
+    const { initialSnapshot, didSnapshotsChange } = options;
+
+    this.subscriptions = new Set();
+    this.activeSnapshot = initialSnapshot;
+    this.didSnapshotsChange = didSnapshotsChange ?? defaultDidSnapshotsChange;
+  }
+
+  private notifySubscriptions(): void {
+    const snapshotBinding = this.activeSnapshot;
+    this.subscriptions.forEach(cb => cb(snapshotBinding));
```

Comment:
Need to make sure that we store the starting snapshot in a separate variable, because the value of `this.activeSnapshot` is allowed to change over time

Otherwise, it's possible for the subscriptions to start getting processed, the snapshot to change halfway through, and then the subscriptions start notifying you about the updated snapshot, not the original one

Reply:
Makes sense.